### ISSUE TITLE
Comment about use of all gather

### ISSUE THREAD

**Issue by NielsRogge** [lucidrains/ring-attention-pytorch#1]
Hi Phil!

Hope you're doing well. As you saw with Gemini Pro 1.5, which works on [1 million tokens](https://blog.google/technology/ai/google-gemini-next-generation-model-february-2024/#context-window), open-source has some work to do to catch up :D porting Ring Attention to PyTorch is definitely the first step towards that.

@rwightman made an interesting comment on your current approach of implementing Ring Attention, tought it would be useful for you to share that: https://twitter.com/wightmanr/status/1758275957557719308. Basically Ross had something similar he had to implement to make the SigLIP loss function work, leveraging neighbour exchange instead of allgather.

Btw, if your implementation is done, I would like to leverage it to port the LWM model that came out 2 days ago (https://github.com/LargeWorldModel/LWM). I would port the model to the Hugging Face Transformers library, by adding a `LWMForCausalLM` class. Since the weights are open-sourced I can convert them to the Transformers format.

Btw are you still active on any Discord channel?

Cheers,

Niels

**Comment by lucidrains** [comment]
hey Niels! good to hear from you and hope you have been well

Ross is reading a repo that is not done yet. i was not planning on using all gather for the ring reduce portion, if that is what he is critiquing.

**Comment by lucidrains** [comment]
@NielsRogge where are you working now Niels?
